src/Pate/Monad/Context.hs

```diff
@@ -92,13 +94,80 @@ funInfoToFunEntry binRepr dfi =
   , PB.functionBinRepr = binRepr
   }
 
+buildParsedBlocks :: MD.DiscoveryFunInfo arch ids -> ParsedBlocks arch
+buildParsedBlocks dfi = ParsedBlocks (Map.elems (dfi ^. MD.parsedBlocks))
+
+-- | Write the macaw CFG for the given discovered function to disk if we are
+-- given a directory to store it in.
+--
+-- Note that this directory is the *base* directory; we use the type repr for
+-- the binary (original vs patched) to ensure that we don't overwrite results
+-- from different binaries on disk.
+saveCFG
+  :: (MM.ArchConstraints arch)
+  => Maybe FilePath
+  -> PBi.WhichBinaryRepr bin
+  -> MD.DiscoveryFunInfo arch ids
+  -> IO ()
+saveCFG mCFGDir repr dfi =
+  F.forM_ mCFGDir $ \cfgDir -> do
+    let baseDir = cfgDir </> show repr
+    SD.createDirectoryIfMissing True baseDir
+    let fname = baseDir </> show (MD.discoveredFunAddr dfi) <.> "cfg"
+    IO.withFile fname IO.WriteMode $ \hdl -> do
+      PPT.hPutDoc hdl (PP.pretty dfi)
+
+-- | Return the basic blocks corresponding to the function that contains the
+-- given basic block
+--
+-- This will return a cached parsed function if possible, or run the macaw
+-- analysis on a new address and cache the result.
+--
+-- This function will return 'Nothing' if the address it is provided (via the
+-- 'PB.ConcreteBlock') is not a mapped address in the binary. As long as the
+-- address is valid, this function will at *least* return a single block with a
+-- macaw translation error.
 parsedFunctionContaining ::
-  MM.ArchConstraints arch =>
+  forall bin arch .
+  (PBi.KnownBinary bin, MM.ArchConstraints arch) =>
   PB.ConcreteBlock arch bin ->
   ParsedFunctionMap arch bin ->
-  Maybe (ParsedBlocks arch)
-parsedFunctionContaining blk (ParsedFunctionMap pfm) =
-  Map.lookup (PB.blockFunctionEntry blk) pfm
+  IO (Maybe (ParsedBlocks arch))
+parsedFunctionContaining blk (ParsedFunctionMap pfmRef mCFGDir) = do
+  let baddr = PA.addrToMemAddr (PB.concreteAddress blk)
+  st <- IORef.readIORef pfmRef
+  let ds1 = discoveryState st
+  let mem = MD.memory ds1
+  -- First, check if we have a cached set of blocks for this state
+  case Map.lookup (PB.blockFunctionEntry blk) (parsedFunctionCache st) of
+    Just blks -> return (Just blks)
+    Nothing -> do
+      -- Otherwise, run code discovery at this address
+      case MM.resolveRegionOff mem (MM.addrBase baddr) (MM.addrOffset baddr) of
+        Nothing -> return Nothing -- This could be an exception... but perhaps better as a proof failure node
+        Just faddr -> do
+          (Some dfi, blks) <- IORef.atomicModifyIORef' pfmRef (atomicAnalysis faddr)
```

Comment:
Do you actually want the strict atomic modify here?  It might be better to allow it to be lazy, as the code discovery process itself is pure.  The lazy version might get you less contention on this IORef.

Reply:
Or... maybe it doesn't matter much, given that the discovery state has to be threaded through.

Reply:
I spent some time trying to reason through what would actually happen here (e.g., if there are retry semantics, and what they are). I'm at "I'm not sure". The returned value is a pair and the value stored in the IORef doesn't have strict fields, so maybe I should just make it lazy to make it clearer that it isn't trying hard to force things to normal form.

Reply:
I will add a note before merging, since I agree that it is unclear and something that might need to be addressed if there ends up being some weird contention that leads to a lot of re-computation